File: polynote-runtime/src/main/scala/polynote/runtime/DataType.scala
package polynote.runtime

sealed trait DataType extends Serializable {
  def size: Int = -1
}

sealed abstract class FixedSizeType(override val size: Int) extends DataType
sealed abstract class PrimitiveType(size: Int) extends FixedSizeType(size)

case object ByteType extends PrimitiveType(1)
case object BoolType extends PrimitiveType(1)
case object ShortType extends PrimitiveType(2)
case object IntType extends PrimitiveType(4)
case object LongType extends PrimitiveType(8)
case object FloatType extends PrimitiveType(4)
case object DoubleType extends PrimitiveType(8)

case object BinaryType extends DataType
case object StringType extends DataType

final case class StructField(
  name: String,
  dataType: DataType
)

final case class StructType(
  fields: List[StructField]
) extends DataType {
  override def size: Int = if (fields.isEmpty || fields.forall(_.dataType.size >= 0)) {
    fields.map(_.dataType.size).sum
  } else -1
}

final case class OptionalType(element: DataType) extends DataType {
  override val size: Int = if (element.size >= 0) element.size + 1 else -1
}

final case class ArrayType(element: DataType) extends DataType
final case class MapType(kv: StructType) extends DataType

//case object DateType extends FixedSizeType(4)
//case object TimestampType extends FixedSizeType(8)

case object TypeType extends DataType

File: polynote-frontend/polynote/data_type.js
'use strict';

import {
    Codec, DataReader, DataWriter, discriminated, combined, arrayCodec,
    str, shortStr, tinyStr, uint8, uint16, int32
} from './codec.js'
import {Result} from "./result";

export class DataType {}
DataType.delegatedCodec = {
    // Need to make a delegator here, because DataType.codec doesn't exist yet and forms recursive knots with i.e. StructField.
    // This defers evaluating DataType.codec, tying the knot
    encode: (value, writer) => DataType.codec.encode(value.dataType, writer),
    decode: (reader) => DataType.codec.decode(reader)
};

function SingletonDataType(msgTypeId) {
    const inst = {
        msgTypeId,
        codec: {
            encode: (value, writer) => {},
            decode: (reader) => inst,
        }
    };
    Object.setPrototypeOf(inst, DataType);
    inst.constructor = inst;
    return inst;
}

export const ByteType = SingletonDataType(0);
export const BoolType = SingletonDataType(1);
export const ShortType = SingletonDataType(2);
export const IntType = SingletonDataType(3);
export const LongType = SingletonDataType(4);
export const FloatType = SingletonDataType(5);
export const DoubleType = SingletonDataType(6);
export const BinaryType = SingletonDataType(7);
export const StringType = SingletonDataType(8);

export class StructField {
    constructor(name, dataType) {
        this.name = name;
        this.dataType = dataType;
        Object.freeze(this);
    }
}

StructField.codec = combined(
    str,
    DataType.delegatedCodec
).to(StructField);

export class StructType extends DataType {
    static get msgTypeId() { return 9; }
    static unapply(inst) {
        return [inst.fields];
    }

    constructor(fields) {
        super(fields);
        this.fields = fields;
        Object.freeze(this);
    }
}

StructType.codec = combined(arrayCodec(int32, StructField.codec)).to(StructType);

export class OptionalType extends DataType {
    static get msgTypeId() { return 10; }
    static unapply(inst) {
        return [inst.element];
    }

    constructor(element) {
        super(element);
        this.element = element;
        Object.freeze(this);
    }
}

OptionalType.codec = combined(DataType.delegatedCodec).to(OptionalType);

export class ArrayType extends DataType {
    static get msgTypeId() { return 11; }
    static unapply(inst) {
        return [inst.element];
    }

    constructor(element) {
        super(element);
        this.element = element;
        Object.freeze(this);
    }
}

ArrayType.codec = combined(DataType.delegatedCodec).to(ArrayType);

export const DateType = SingletonDataType(12);
export const TimestampType = SingletonDataType(13);

DataType.codecs = [
    ByteType,     //  0
    BoolType,     //  1
    ShortType,    //  2
    IntType,      //  3
    LongType,     //  4
    FloatType,    //  5
    DoubleType,   //  6
    BinaryType,   //  7
    StringType,   //  8
    StructType,   //  9
    OptionalType, // 10
    ArrayType,    // 11
    DateType,     // 12
    TimestampType, // 13
];

DataType.codec = discriminated(
    uint8,
    (msgTypeId) => DataType.codecs[msgTypeId].codec,
    (result) => result.constructor.msgTypeId
);